### PR TITLE
Block closing the reader until the current read has unblocked

### PR DESCRIPTION
**Pull request by dmathieu**
`Readers` are blocking, making it so that when we get a disconnection, the reader is still listening on stdin until it gets some data.
`#Close` isn't blocking though. So we think we closed the connection and can move on to the next one, when we're still reading for data in a connection which finished.

This mutex makes `#Close` blocking, so closing the body waits until the next bit has been received and written to the buffer.
The next connection can then kick in and not miss any data.

### CODE REVIEW

**Comment by tt** [comment]
What if the `Read` hangs? Isn't it an important property to be able to always close the connection?

What's the consequence of closing it without the synchronization? Shouldn't that itself end the `Read` call?

**Comment by dmathieu** [comment]
The thing is, if `Read` hangs, it also blocks, and will wait until it receives a byte.
The effect is that we close the connection. But are still waiting on some data to read, which will be buffered
Then, as there's already something listening on stdin, the next connection doesn't get any data.

This mutex causes `Close` to block until `Read` has finished blocking, meaning we don't actually open a new connection and don't start listening on stdin.
Then the new connection only opens once the previous one isn't reading anymore, and properly gets all data from the buffer and the stream.

File: busltee/transport_test.go [hunk 1]
```diff
@@ -128,3 +129,40 @@ func TestHTTPError(t *testing.T) {
 		t.Fatalf("was expecting 9 retries. Got %d", callCount)
 	}
 }
+
+type slowBuffer struct {
+}
```
Comment (tt) [inline busltee/transport_test.go]: Can you move this to the previous line?

File: busltee/transport_test.go [hunk 2]
```diff
@@ -128,3 +129,40 @@ func TestHTTPError(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer buffer.Close()
+
+	reader, err := newBodyReader(streamer, buffer)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	go func() {
```
Comment (tt) [inline busltee/transport_test.go]: Maybe just `go reader.Close()`?

File: busltee/transport_test.go (hunk 1 above)
Comment (dmathieu) [inline busltee/transport_test.go]: :+1:

File: busltee/transport_test.go (hunk 2 above)
Comment (dmathieu) [inline busltee/transport_test.go]: Nice catch. Done.

**Pull request merged**